src/Microsoft.DotNet.VersionTools/BuildManifest/BuildManifestClient.cs

```diff
@@ -0,0 +1,214 @@
+
+                XElement modifiedModelXml = modifiedModel.ToXml();
+
+                IEnumerable<SupplementaryUploadRequest> uploads = semaphorePaths.NullAsEmpty()
+                    .Select(p => new SupplementaryUploadRequest
+                    {
+                        Path = p,
+                        Contents = new SemaphoreModel
+                        {
+                            BuildId = orchestratedBuildId
+                        }.ToFileContent()
+                    })
+                    .Concat(supplementaryUploads.NullAsEmpty())
+                    .ToArray();
+
+                if (!XNode.DeepEquals(modifiedModelXml, remoteModelXml))
+                {
+                    uploads = uploads.Concat(new[]
+                    {
+                        new SupplementaryUploadRequest
+                        {
+                            Path = BuildManifestXmlName,
+                            Contents = modifiedModelXml.ToString()
+                        }
+                    });
+                }
+
+                return await PushUploadsAsync(project, @ref, basePath, message, remoteCommit, uploads);
+            });
+        }
+
+        private async Task<XElement> FetchModelXmlAsync(
+            GitHubProject project,
+            string @ref,
+            string basePath)
+        {
+            string contents = await _github.GetGitHubFileContentsAsync(
+                $"{basePath}/{BuildManifestXmlName}",
+                project,
+                @ref);
+
+            return XElement.Parse(contents);
+        }
+
+        private async Task<bool> PushUploadsAsync(
+            GitHubProject project,
+            string @ref,
+            string basePath,
+            string message,
+            string remoteCommit,
+            IEnumerable<SupplementaryUploadRequest> uploads)
+        {
+            GitObject[] objects = uploads
+                .Select(upload => new GitObject
+                {
+                    Path = $"{basePath}/{upload.Path}",
+                    Mode = GitObject.ModeFile,
+                    Type = GitObject.TypeBlob,
+                    Content = upload.Contents
+                })
+                .ToArray();
+
+            GitTree tree = await _github.PostTreeAsync(project, remoteCommit, objects);
+
+            GitCommit commit = await _github.PostCommitAsync(
+                project,
+                message,
+                tree.Sha,
+                new[] { remoteCommit });
+
+            try
+            {
+                // Only fast-forward. Don't overwrite other changes: throw exception instead.
+                await _github.PatchReferenceAsync(project, @ref, commit.Sha, force: false);
+            }
+            catch (NotFastForwardUpdateException e)
+            {
+                // Retry if there has been a commit since this update attempt started.
+                Trace.TraceInformation($"Retrying: {e.Message}");
+                return false;
```

Comment:
If we can't fast forward here, will the exponential retry run instead?  Would that mean we would retry even if it was always going to fail?

Reply:
Yeah,
`true` => success, don't run again,
`false` => run the Func again,
`throw` => abort and let the exception propagate. (Func is responsible for catching.)

It should be impossible for this to always fail--the window for a `NotFastForwardUpdateException` is very short. Someone else has to push a commit between the `GetReferenceAsync` call and this `PatchReferenceAsync` call. The retry does both calls again, so there's another small window that happens immediately after the first one since the first retry is instant.

If we start getting this exception for a different reason, that would be an issue... the detection is very specific, though.